Implements In App Account Deletion

danilo04 opened the pull request:
Fixes #1415 

### Fix
Implement a feature that let users delete their account from within the app. This feature sends a request to the Simplenote server which sends an email to the user. The user can then delete the account using the email with the required information. If the request fails because of Internet connection or server issues, then we show an error message to the user, otherwise, we show a received request message. The videos below show both flows:

**Light Mode**
| Successful Flow      | Error Flow |
| ----------- | ----------- |
| ![20210716_001406](https://user-images.githubusercontent.com/195721/125895508-44d36a29-ac0b-4366-bf65-fd45f3ae01c4.gif) | ![20210716_001512](https://user-images.githubusercontent.com/195721/125895527-7213c035-64db-4e98-909e-9e65e6a1985e.gif)      |


**Dark Mode**
| Successful Flow      | Error Flow |
| ----------- | ----------- |
| ![20210716_001948](https://user-images.githubusercontent.com/195721/125895958-8e8ea1a4-e87b-4aa6-ab63-0e3bad6d89f8.gif)  |  ![20210716_002042](https://user-images.githubusercontent.com/195721/125895979-1c159760-04a1-4dde-bf66-6c0b6c851854.gif)    |


More details about the project: p2XJRt-3bK-p2
Last Iteration of designs: p2XJRt-3cd-p2


### Test

**Note: use Simperium key for production and create accounts for tests. You can create accounts with the following format: youremail+n@automattic.com where n is an integer number (e.g. danilo.dominguez+1@automattic.com). Follow the instructions in the email to setup password.**

1. Open the left menu by tapping on the top-left corner icon
2. Tap on Settings
3. Scroll down and Tap on the preference `Delete Account`
4. That will open a dialog. Tap on `DELETE ACCOUNT` button
5. Open your email and proceed to delete account
6. Open Simplenote and it should logout automatically


### Review

This PR can be reviewed by @ParaskP7  or @hichamboushaba.

### Release
`RELEASE-NOTES.txt` was updated in fd9ac10108a556a42ce1b2dbe33c3babb63d90d5 with:
> Added In App Account Deletion

danilo04 committed:
Added button Delete Account in Preferences screen

Added a new layout to handle red text preferences to be use in the
Delete Account button. Plus, added the placeholder for the onClick
listener

danilo04 committed:
Added dialogs to handle account deletion

Added the first dialog after clicking on the button to delete account
and the dialog that is shown when the account deletion request was
sent successfully.

danilo04 committed:
Added dialogs to handle error message in case of unsuccessful call

In case a call to delete an account is unsuccessful, an error message
is shown in a dialog following the pattern used in
simplenote-android for error messages.

danilo04 committed:
Added helper to handle account deletion HTTP requests

Util class that can be use to handle HTTP requests for account deletion
and can be extended to handle other requests such as signing up.

danilo04 committed:
Added action to make http request to delete account inside click handler

Added call to make http request to delete account in the event handler
of the Delete Account preference. This commit also includes the
definition of the callback that handles the response of the call.
Depending on whether the call was successful or not, we show a dialog
accordingly.

danilo04 committed:
Fix endpoint for account deletion

danilo04 committed:
Run UI updates inside delete account callback on the UI thread

Since the callback that receives the response from an OkHttpClient call
runs on a background thread, the calls that show the different dialogs
to handle a successful or an error response should be run on the UI
thread.

danilo04 committed:
Removed unneeded final in method

danilo04 committed:
Added progress bar (spinner) to the delete account request

When the user tabs on delete account button, the app sends a HTTP
request to the Simplenote server. This operation can have some latency
depending on the server and the user's Internet connection. Thus, we
show a visual cue to the user that a request is being made.

danilo04 committed:
Changed HTTTP host for account deletion to a temporary host for testing

danilo04 committed:
Changed message in spinner while requesting to delete account

Changed the message showed next to the spinner to "Requesting..." while
waiting for the response of the delete account request.

danilo04 committed:
Changed message in dialog showed after a successful account deletion

Changed message showed in dialog after a successful response from an
account deletion request. The new message follow the new design changes.

danilo04 committed:
Changed mechanism to get color in a fragment, previous one is deprecated

danilo04 committed:
Added checks to avoid NPE and crashes while delete account actions

Avoid NPE when the activity class is required. Also added try catch in
methods that could throw exceptions in edge cases.

danilo04 committed:
Added AppLog logs to delete account process

danilo04 committed:
Use ellipsis unicode for strings used in requesting delete account

danilo04 committed:
Check for nullability when getting context to create dialogs

Instead of requireContext(), use getContext() and check whether the
context is null. If it is null, it means that activity of the fragment
was destroyed.

danilo04 committed:
Added In App Account Deletion to release notes

danilo04 committed:
Added a custom progress dialog fragment to handle different themes

The ProgressDialogFragment class that was used for progress bar dialogs
(part of Simperium library and used in the login) does not support dark
mode. When the dark mode was activated, the progress bar showed in
white. We implemented a custom layout which is used with an AlertDialog
to implement the same behavior as ProgressDialog but with support to
dark mode and other themes.

danilo04 committed:
Removed unneeded style on the progress dialog for deleting account

SylvesterWilmott commented:
Thanks @danilo04 

This looks great and works as expected. Design-wise theres just a few things:

- In dark mode, the red text in the buttons in the settings and dialog should be red20 to provide adequate contrast. This is the general rule we use in dark mode ie. all text should use the `20` variant of the light mode's counterpart.
- Can we update the copy of the error copy to - `An error occurred. Please, try again. If the problem continues, contact us at support@simplenote.com for help.`
- I noticed that in the "Request Received" dialog, my email address was broken (ie. it flowed onto two lines), is this a consequence of it being too long to fit on one line or do we need to make sure that the dialog doesn't break this particular word?

danilo04 committed:
Updated error message for dialog in delete account response

danilo04 committed:
Changed styles of red for dark mode

For light mode, the red color used is red_50. For dark mode the red
color that should be used is red_20. We modified the preference layouts
and added a layout of preference red in light and another one for
preference red in dark mode.

danilo04 committed:
Fixed summary color for preferences' layout in red

danilo04 commented:
Hi @SylvesterWilmott, thank you for the review. 

> In dark mode, the red text in the buttons in the settings and dialog should be red20 to provide adequate contrast. This is the general rule we use in dark mode ie. all text should use the 20 variant of the light mode's counterpart.

Fixed in 572d4dd46cc26b211b7fabee210c777d67b82de8 and 2bd355e31edae1ac5f486a1f38b4c673fbc5f6de. 

> Can we update the copy of the error copy to - An error occurred. Please, try again. If the problem continues, contact us at support@simplenote.com for help.

Fixed in c2aa7aa5345a218fa7bf33727771a1cdb07772de.

> I noticed that in the "Request Received" dialog, my email address was broken, is this a consequence of it being too long to fit on one line or do we need to make sure that the dialog doesn't break this particular word?

Let me investigate this case because it is breaking just the email in any case. If you rotate your phone, the email can be put in the next line, but it is still breaking it.

hichamboushaba commented on Simplenote/src/main/java/com/automattic/simplenote/PreferencesFragment.java:
```diff
@@ -116,6 +130,18 @@ public boolean onPreferenceClick(Preference preference) {
         });
 
+        Preference deleteAppPreference = findPreference("pref_key_delete_account");
+        int preferenceLayout = ThemeUtils.isLightTheme(requireActivity()) ?
+                R.layout.preference_red_light : R.layout.preference_red_dark;
```
Just a suggestion, I didn't test it to see if it works, why not specify the layout directly in `preferences.xml` using the `android:layout` option, to have a separate layout for dark them maybe you can define it in the `layout-night` folder, or define the colors in the `values-night` folder, and use the same layout.
WDYT?

danilo04 replied:
Yes, I think that makes the code much cleaner and removes code from the fragment that should be in the layout. It also makes it simpler and flexible to add more themes to the app. Fixed in c177fab7e3144f5469c27ed2a048105783961dcc.

hichamboushaba commented on Simplenote/src/main/java/com/automattic/simplenote/PreferencesFragment.java:
```diff
@@ -304,6 +330,185 @@ public boolean onPreferenceClick(Preference preference) {
+
+                AppLog.add(Type.ACCOUNT, "Failure while calling server to delete account");
+
+                activity.runOnUiThread(new Runnable() {
+                    @Override
+                    public void run() {
+                        closeProgressDialogDeleteAccount();
+                        showDialogDeleteAccountError();
+                    }
+                });
+            }
+
+            @Override
+            public void onResponse(Call call, final Response response) throws IOException {
+                FragmentActivity activity = getActivity();
+                if (activity == null) {
+                    return;
+                }
+
+                activity.runOnUiThread(new Runnable() {
+                    @Override
+                    public void run() {
+                        closeProgressDialogDeleteAccount();
+
+                        // The delete account requests return 200 when the request was processed
+                        // successfully. These requests send an email to the user with instructions
+                        // to delete the account. This email is valid for 24h. If the user sends
+                        // another request for deletion and the previous request is still valid,
+                        // the server sends a response with code 202. We take both 200 and 202 as
+                        // successful responses. Both codes are handled by isSuccessful()
+                        if (response.isSuccessful()) {
+                            AppLog.add(Type.ACCOUNT, "Request to delete account was successful");
+
+                            showDeleteAccountConfirmationDialog();
+                        } else {
+                            AppLog.add(Type.ACCOUNT, "Request to delete account had an " +
+                                    "error. Error code: " + response.code());
+
+                            showDialogDeleteAccountError();
+                        }
```
just a thought, I don't think it's a good idea to expose OkHttp's CallBack here, it's making the fragment responsible for network logic (response.isSuccessFull).
I would prefer to have some kind of interface as listener for the account deletion, with two simple methods: `onSuccess` and `onFailure`, which we can then pass to `makeDeleteAccountRequest`, and in `AccountNetworkUtils` we would use the `Callback` to handle the logic of success and failure and forward it to our listener.

danilo04 replied:
HI @hichamboushaba 👋 . Yes, I agree, it is not the responsibility of the fragment. I have added a handler in 28c1446a4d80a59936bc5de194bd53560cc0a070 and move the fragment to a weak reference in the handler (dd8199e4f88400f2d43936ddc55bfbeb5f1d887c) to avoid memory leaks.

hichamboushaba reviewed (approved):
Good job @danilo04, the code looks code, and the change works as expected.

I left some comments for you to consider, but I'm pre-approving as they are not blockers.

danilo04 committed:
Set layout of delete account preference in XML file

Instead of setting the layout of the delete account preference in code,
we set the layout in the XML file. Moving this to the layout keep the
fragment from doing something the should be done in the layout. We also
simplify the layout and added just one layout for light and dark mode
(preferences_red.xml). This layout uses a user-defined color that is set
to red_50 for light and red_20 for dark themes.

danilo04 committed:
Moved network logic from fragment for account deletion

Moved network logic defined in a OkHttp.Callback from
PreferencesFragment to AccountNetworkUtils so it takes care of handling
the response. makeDeleteAccountRequest now uses an interface which does
not have the responsibility of dealing with network logic.

danilo04 committed:
Added a weak reference to the fragment in the delete account handler

Changed DeleteAccountRequestHandlerImpl to a static class and added a
weak reference to PreferencesFragment to avoid memory leaks in case the
fragment is destroyed before the OkHttp callback returns.

danilo04 committed:
Merge branch 'develop' into issue/1415-in-app-account-deletion

danilo04 committed:
Added tracks to delete account request

danilo04 commented:
Hi @SylvesterWilmott 👋 . I have added track events for delete account in cc253a776c4772647e72f7afdcabd389d6ad9a17.

danilo04 committed:
Merge branch 'develop' into issue/1415-in-app-account-deletion

danilo04 committed:
Merge branch 'develop' into issue/1415-in-app-account-deletion

SylvesterWilmott reviewed (approved):
Approving the design on this since I believe the email breaking will be tackled separately. Thanks @danilo04

danilo04 committed:
Changed test URL to production URL for account deletion request

danilo04 commented:
I am going to merge the PR since it has been approved.
@SylvesterWilmott, I have added the issue regarding how the email is shown in two lines: #1427.
